File: sample/flutter_base/packges/service_manager/lib/src/ServiceManager.dart
abstract class ServiceProtocol {
  void onLoad();
  void onUnLoad();
}

class ServiceManager {
  final hashMap = Map<String, ServiceProtocol>();

  static late final ServiceManager sInstance = ServiceManager();
  static ServiceManager getInstance() => sInstance;

  void setProtocol(String name, ServiceProtocol component) {
    this.hashMap[name] = component;
  }

  ServiceProtocol getProtocol(String name) {
    final component = this.hashMap[name] as ServiceProtocol;
    return component;
  }
}

File: output/src/services/ServiceManager.ts
// src/services/ServiceManager.ts

export interface ServiceProtocol {
  onLoad(): void;
  onUnLoad(): void;
}

class ServiceManager {
  private hashMap: Record<string, ServiceProtocol> = {};

  private static sInstance: ServiceManager | null = null;

  static getInstance(): ServiceManager {
    if (!ServiceManager.sInstance) {
      ServiceManager.sInstance = new ServiceManager();
    }
    return ServiceManager.sInstance;
  }

  setProtocol(name: string, component: ServiceProtocol): void {
    this.hashMap[name] = component;
  }

  getProtocol(name: string): ServiceProtocol {
    // Dart's `as ServiceProtocol` cast is a no-op in TypeScript; 
    // returning the value ensures type safety.
    return this.hashMap[name]!;
  }
}

export { ServiceManager };